emcc.py

```diff
@@ -516,6 +515,44 @@ def get_all_js_syms():
     if shared.is_c_symbol(name):
       name = shared.demangle_c_symbol_name(name)
       library_syms.add(name)
+  return library_syms
+
+
+@ToolchainProfiler.profile_block('JS symbol generation')
+def get_all_js_syms():
+  # To avoid the cost of calling generate_js_symbols each time an executable is
+  # linked we cache symbol lists for the N most recently used configs.
+  # We define a cache hit as when the settings and `--js-library` contents are
+  # identical.
+  input_data = json.dumps(settings.dict(), sort_keys=True, indent=2) + '\n'
+  for jslib in settings.JS_LIBRARIES:
+    if os.path.abspath(jslib):
+      jslib = utils.path_from_root('src', jslib)
+    input_data += read_file(jslib)
+  cache_filename = None
+  num_cache_entries = 20
+
+  with cache.lock('js_symbol_lists'):
+    oldest_timestamp = 0
+    for i in range(num_cache_entries):
```

Comment:
I think it would be good to use hashing to avoid looping on all existing cache entries - we just need to check the one with the same hash. Otherwise this is linear in the number of entries. In particular in our test suite I imagine many runs with have unique js libraries and we could end up with hundreds or thousands of items we check in each test run.

Reply:
But the number of entries is constant (20)..  how would do that hashing?   Would do you imagine doing the hashing?   Would you somehow put the hash into the filename maybe?